net: lib: http_server: Close sockets on server core errors

rlubos opened the pull request:
In case of fatal errors (during poll() or when handling listening socket), the server operation is restarted. It was missed however, that sockets opened for the server should be closed in such case. This i commit fixes it.

Fixes #75658

rlubos committed:
net: lib: http_server: Close sockets on server core errors

In case of fatal errors (during poll() or when handling listening
socket), the server operation is restarted. It was missed however, that
sockets opened for the server should be closed in such case.

Additionally, in case there were active client connections, it's needed
to cleanup related resources, otherwise running timers may trigger a
crash.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
DNM for now, the fix is certainly valid but apparently incomplete to resolve the linked issue - need to investigate this more.

rlubos committed:
net: lib: http_server: Add restart delay

In case there were active connections when restarting the server, it
can't be re-initialized immediately, as binding to the server port will
fail. We need to wait for the TCP connection teardown, as even with
REUSEADDR socket option set, binding will fail if the sockets are not in
TCP TIMED_WAIT state (i. e. connections are active).

Because of this, add a configurable delay when restarting the server.
Additionally, make server initialization failures non-fatal, i. e. try
to restart the server again after the delay if the initialization fails.
It's been observed with Chromium, that it tends to keep connections open
even after closing them on the server side (socket lingers in FIN_WAIT_2
state), so the server re-initialization may fail even with delay, so
it's beneficial in  such case to keep retrying the server
re-initialization.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: sockets: Allocate FD after receiving connection in accept()

Allocating FD before pulling the new connection from the fifo can lead
to busy looping in certain cases. If the application keeps calling
accept() on a listening socket after failing to allocate new FD for the
incoming connection, it'll start busy looping, as will report POLLIN in
such case (as the new connection is still on the queue), but it'll
consistently fail with ENFILE.

This can be avoided by trying to allocate new file descriptor only after
new connection has been pulled from the fifo. That way, if we fail to
allocate the file descriptor, the incoming connection will be dropped,
which seems correct given we don't have enough resources to service it.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

pdgendt commented on subsys/net/lib/http/Kconfig:
```diff
@@ -123,6 +123,16 @@ config HTTP_SERVER_RESOURCE_WILDCARD
+config HTTP_SERVER_RESTART_DELAY
+	int "Delay before re-initialization when restarting server"
+	default 1000
+	range 0 60000
```
This still feels error-prone as I read the description, would 0 ever be desired?

rlubos replied:
Possibly not, although with this range I just wanted to ensure no one sets a negative delay. I want to be able to set a very small delay here, to verify the feature with ztest (I'm working on extending the server test coverage).

rlubos replied:
On second thought, I've changed the lower range to 1, just in case. Setting 0 could indeed have nasty consequences if initialization fails. Thanks for pointing that out.

rlubos committed:
net: lib: http_server: Respect Connection close header

In case no "Connection: close" header is present in the request, the
server should keep the connection open for the client. Hence, after
serving a request, we need to check if the header was present (the
parser sets a flag for it), and only close the connection immediately,
if the client requested it. In case the client remains silent, the
connection will be closed anyway after the inactivity timer kicks in.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
This stats to get bulky... Added another commit, fixing the behavior when there's not Connection: Close header in HTTP1 request. The connection should remain open is such case to handle further requests.
